### PR TITLE
TypeScript: views & viewports

### PR DESCRIPTION
**Pull request by Pessimistress**
For https://github.com/visgl/deck.gl/issues/5589

#### Change List
- Bump math.gl to 3.6.0-alpha (typed)
- Viewport classes
- View classes
- ViewManager

### CODE REVIEW

File: examples/website/mesh/package.json [hunk 1]
```diff
@@ -8,7 +8,7 @@
   },
   "dependencies": {
     "@loaders.gl/obj": "^3.1.5",
-    "@math.gl/core": "^3.5.6",
+    "@math.gl/core": "3.6.0-alpha.1",
```
Comment (ibgreen) [inline examples/website/mesh/package.json:11]: Different alphas (.1 vs .3)...

File: modules/core/src/controllers/controller.ts [hunk 2]
```diff
@@ -45,30 +45,50 @@ const EVENT_TYPES = {
...
+        rotateSpeedY?: number;
       };
+  /** Drag behavior without pressing function keys, one of `pan` and `rotate`. */
   dragMode?: 'pan' | 'rotate';
+  /** Enable inertia after panning/pinching. If a number is provided, indicates the duration of time over which the velocity reduces to zero, in milliseconds. Default `false`. */
   inertia?: boolean | number;
-} & TransitionProps;
+};
+
+type ControllerProps = {
```
Comment (ibgreen) [inline modules/core/src/controllers/controller.ts:85]: I vote for adding JSDoc on every type - the comments are presented by vscode.

File: modules/core/src/controllers/transition-manager.ts [hunk 3]
```diff
@@ -14,12 +14,19 @@ export enum TRANSITION_EVENTS {
 }
 
 export type TransitionProps = {
+  /** Transition duration in milliseconds, default value 0, implies no transition. When using `FlyToInterpolator`, it can also be set to `'auto'`. */
```
Comment (ibgreen) [inline modules/core/src/controllers/transition-manager.ts:17]: Add JSDoc on type itself.

File: modules/core/src/index.ts [hunk 4]
```diff
@@ -94,3 +97,11 @@ export {count as _count} from './utils/count';
 export {default as _memoize} from './utils/memoize';
 export {mergeShaders as _mergeShaders} from './utils/shader';
 export {compareProps as _compareProps} from './lifecycle/props';
+
+/* Types */
```
Comment (ibgreen) [inline modules/core/src/index.ts:101]: Nit: Consistency?
```suggestion
// Types
```

File: modules/core/src/index.ts [hunk 5]
```diff
@@ -53,6 +53,9 @@ export {default as DeckRenderer} from './lib/deck-renderer';
 export {default as Viewport} from './viewports/viewport';
 export {default as WebMercatorViewport} from './viewports/web-mercator-viewport';
 export {default as _GlobeViewport} from './viewports/globe-viewport';
+export {default as OrbitViewport} from './viewports/orbit-viewport';
```
Comment (ibgreen) [inline modules/core/src/index.ts:56]: Why expose these? There is a lot going on in this PR...

File: modules/core/src/lib/view-manager.ts [hunk 6]
```diff
@@ -50,17 +87,18 @@ export default class ViewManager {
...
-        this.controllers[key].finalize();
+      const controller = this.controllers[key];
+      if (controller) {
+        controller.finalize();
       }
     }
     this.controllers = {};
   }
 
   // Check if a redraw is needed
```
Comment (ibgreen) [inline modules/core/src/lib/view-manager.ts:100]: Convert method comments to proper JSDoc so vscode picks them up...

File: modules/core/src/lib/view-manager.ts [hunk 7]
```diff
@@ -70,13 +108,13 @@ export default class ViewManager {
 
   // Layers will be updated deeply (in next animation frame)
   // Potentially regenerating attributes and sub layers
-  setNeedsUpdate(reason) {
+  setNeedsUpdate(reason: string) {
```
Comment (ibgreen) [inline modules/core/src/lib/view-manager.ts:111]:
```suggestion
  setNeedsUpdate(reason: string): void {
```

**Review by ibgreen — APPROVED** [review]
Nice. A big step forward for the core module!

File: modules/core/src/controllers/controller.ts [hunk 8]
```diff
@@ -45,30 +45,50 @@ const EVENT_TYPES = {
...
 
-export type ControllerProps = {
-  id: string;
-  x: number;
-  y: number;
-  width: number;
-  height: number;
-
-  scrollZoom?: boolean | {speed?: number; smooth?: boolean};
+export type ControllerOptions = {
```
[the remark this answers is not in the document]
Comment (Pessimistress) [reply modules/core/src/controllers/controller.ts:48]: The controller is never explicitly instantiated by the user. Part of the controller props are defined in the View, and part is dynamically set in the view state.

**Pull request merged**